app/models/claim/scrubber.rb

```diff
@@ -0,0 +1,47 @@
+# Removes attributes from a claim and its amendments
+class Claim
+  class Scrubber
+    def self.scrub!(claim, attributes_to_delete)
+      new(claim, attributes_to_delete).scrub!
+    end
+
+    attr_reader :claim, :attributes_to_delete
+
+    def initialize(claim, attributes_to_delete)
+      @claim = claim
+      @attributes_to_delete = attributes_to_delete.map(&:to_s)
+    end
+
+    def scrub!
+      ApplicationRecord.transaction do
+        claim.amendments.each { |amendment| scrub_amendment!(amendment) }
+        scrub_claim!
+        scrub_session!
+      end
+    end
+
+    private
+
+    def scrub_amendment!(amendment)
+      amendment_data_to_scrub = attributes_to_delete & amendment.claim_changes.keys.map(&:to_s)
+      personal_data_mask = amendment_data_to_scrub.to_h { |attr| [attr, nil] }
+      amendment.claim_changes.merge!(personal_data_mask)
+      amendment.personal_data_removed_at = Time.zone.now
+      amendment.save!
+    end
+
+    def scrub_claim!
+      personal_data_mask = attributes_to_delete.to_h { |attr| [attr, nil] }
+      attributes_to_set = personal_data_mask.merge(
+        personal_data_removed_at: Time.zone.now
+      )
+      claim.update_columns(attributes_to_set)
+    end
```

Comment:
why do we call `save!` on the amendment but `update_columns` on the claim.
so `save!` will fire validation and callbacks but `update_columns` will not
is there a reason for the difference? or should they both be calling the same thing?

Reply:
We use `update_columns` to skip any callbacks and validations. The previous version of this used `update_all` for persisting the changes to the claim and `save!` for persisting the changes to the amendment so I wanted to preserve that behaviour.

Reply:
Using `save!` here will mean the callbacks on the claim fail, as there are a bunch of `normalise_*` callbacks that assume those fields are populated.

Reply:
It would be preferable to use `save` because the [dfe_analytics](https://github.com/DFE-Digital/dfe-analytics) gem  does not hook into `update_all` or `update_columns`. This causes data inconsistencies in the BigQuery database (although PII is not sent there in the first place).

Reply:
I don't think we can use `save` due to the callbacks Alkesh highlighted. Does the existing data scrubbing implementation create data inconsistency issues in big query? We could try using `save` in a suppress callbacks block.

Reply:
could we update the 3 affected callbacks (for `ni_number`, `first_name` and `surname` so they don't get called if that field is nil?
e.g.
`before_save :normalise_ni_number, if: [:national_insurance_number, :national_insurance_number_changed?]`

Reply:
> could we update the 3 affected callbacks (for `ni_number`, `first_name` and `surname` so they don't get called if that field is nil? e.g. `before_save :normalise_ni_number, if: [:national_insurance_number, :national_insurance_number_changed?]`

That seems to do the trick 👍